crates/sui-core/src/authority/authority_store.rs

```diff
@@ -234,6 +234,58 @@ impl<S: Eq + Debug + Serialize + for<'de> Deserialize<'de>> SuiDataStore<S> {
             .map(|c| c.into()))
     }
 
+    pub fn store_pending_certs_and_effects(
+        &self,
+        certs: &[VerifiedCertificate],
+        effects: &[ValidTransactionEffects],
+        execution_digests: &[ValidExecutionDigests],
+    ) -> SuiResult {
+        let existing_effects = self
+            .epoch_tables()
+            .valid_effects
+            .multi_get(execution_digests.iter().map(|digests| digests.transaction))?;
+
+        // Only insert digests if there isn't a pre-existing key, so that we don't downgrade
```

Comment:
How can we be sure this is guaranteed under concurrent writes & reads to the two tables?

Reply:
very good catch, thanks